### ISSUE TITLE
Upload wizard: Remove "Please wait" if possible and do the task in background

### ISSUE THREAD

**Issue by nicolas-raoul** [commons-app/apps-android-commons#4704]
Making the user wait is never great.
Is this wait really indispensible?
Couldn't we perform the same tasks in the background?
![Screenshot_20211114-192943_Commons](https://user-images.githubusercontent.com/99590/141683417-4cdc1e0d-cb61-4fd3-929b-adfcdd983878.jpg)

**Comment by misaochan** [comment]
AFAIK this is to check certain things that would impact the ability to upload at all (or the user's choice to upload), e.g. duplicate images and dark images. It may be possible to handle this in the background, but in that case the user would have started typing their caption and desc, only to be told that this image is a duplicate of one that already exists?

**Comment by nicolas-raoul** [comment]
Thanks for the explanation!
So I guess the wait is unavoidable **for the first picture**.

How about for the second/third/etc pictures? :-)
Ideally the app would be able to detect darkness/duplicates/etc of picture N while the user is typing the caption of picture N-1?

Actually, I often have 20 pictures of the same thing under various angles, and just use the same caption for all of them. In that context, I would like to be able to press Next/Next/Next/... very quickly after entering a caption only for the first picture and pressing "Copy to subsequent media". So, ideally the darkness/duplicates/etc detections should all be queued and performed as soon as possible, which means that for instance all 20 pictures might have been processed already by the time I finish writing the caption.

**Comment by misaochan** [comment]
I agree, that would be the ideal way to handle it. :)

**Comment by ShashwatKedia** [comment]
@nicolas-raoul I would like to work on this :)

**Comment by ShashwatKedia** [comment]
> AFAIK this is to check certain things that would impact the ability to upload at all (or the user's choice to upload), e.g. duplicate images and dark images. It may be possible to handle this in the background, but in that case the user would have started typing their caption and desc, only to be told that this image is a duplicate of one that already exists?

Currently, the checks are performed only when the user clicks on the next button, which means that the user gets to know that image is a duplicate (or too dark or downloaded from internet, etc) only after typing the caption and desc (since the next button is disabled without the caption)

**Comment by ShashwatKedia** [comment]
I think the ideal flow could be:
1. As soon as the user uploads an image, first, the checks of only the 1st image are to be performed (meanwhile user is shown some explanation about why the progress bar is shown, instead of just a "please wait..").
2. If all checks are passed, then the user can enter the caption and desc and click NEXT, while in the background, the checks for the next images are performed. Otherwise, problems with the current image are shown.
3. On the NEXT button click, we check if the location for the image is added or not, and display a dialog if it is not added.
4. Then, when the user skips/adds location and the next image's details are shown, we fetch the check results of that image and show to the user if any problems exist. 
5. If no problem exists (or user chooses to continue upload anyways), repeat step 3.

What are your thoughts on this @nicolas-raoul  @misaochan

**Comment by ShashwatKedia** [comment]
@nicolas-raoul I worked on this, and after a lot of changes to the original logic, I could make [this](https://github.com/commons-app/apps-android-commons/assets/142137555/c7e45797-b00a-4548-9125-e5269f2781e7).  Could you confirm that this behaviour is what this issue aims for, so that I could proceed with the minor changes left to be implemented.

**Comment by ShashwatKedia** [comment]
@nicolas-raoul I have worked on this issue, and after a lot of changes to the original logic, I could make the app calculate the quality of the images in the background and show any problems as soon as the images are loaded [[Screencast]](https://github.com/commons-app/apps-android-commons/assets/142137555/c7e45797-b00a-4548-9125-e5269f2781e7). Could you please confirm that this behaviour is what this issue aims for, so that I could proceed with the minor changes left to be implemented or any changes which might be required :)

**Comment by ShashwatKedia** [comment]
@nicolas-raoul pinging you just in case you missed this issue :)

**Comment by ShashwatKedia** [comment]
@nicolas-raoul @misaochan Since there was no response from either of you, I took the liberty to solve this issue in the best possible way I could think of, hope that is not a problem :) Could you please take a look at the above PR?

**Comment by ShashwatKedia** [comment]
@nicolas-raoul @sivaraam, sorry for pinging again, but there has been no reply/comment on my questions and changes for the past month, so am just confirming if this enhancement is still required :(

**Comment by sivaraam** [comment]
Hi @ShashwatKedia ! Apologies about the delay in response. This would indeed be a worthy enhancement for the app. Thank you for taking the time to work on and improve the user experience of this flow! 🙂 

I took a look at the screencast and the flow looks neat. It seems you've also handled the duplicate file detection that should be done after the user enters information which is good.

I'll try to take a look at your PR soon.

**Comment by misaochan** [comment]
Sorry for the late response, @ShashwatKedia ! I agree that the new flow is better. :)
